odoo/modules/loading.py

```diff
@@ -237,6 +237,21 @@ def load_module_graph(env, graph, status=None, perform_checks=True,
         if package.name is not None:
             registry._init_modules.add(package.name)
 
+        if needs_update:
+            processed_modules.append(package.name)
+
+            ver = adapt_version(package.data['version'])
+            # Set new modules and dependencies
+            module.write({'state': 'installed', 'latest_version': ver})
+
+            package.load_state = package.state
+            package.load_version = package.installed_version
+            package.state = 'installed'
+            for kind in ('init', 'demo', 'update'):
+                if hasattr(package, kind):
+                    delattr(package, kind)
+            module.env.flush_all()
```

Comment:
Moving this code here makes sense, but I would move just before the `env.cr.commit()` below instead.
So that changing the module's state would be the last thing done in the transaction.